[#2] 포스트와 함께 구성된 프로필 구현

khoon9 opened the pull request:
# 구현 API
- 프로필
    - get 
        - 프로필 수정
        - 프로필 조회(일대다 관계 포스트 내용 포함)
- 포스트
    - post
        -  포스트 생성
    - get
        -  포스트 조회
    - delete
        -  포스트 삭제

# 유의
- 포스트 entity의 이미지 처리를 위해 AWS SDK v2와 S3 client 의존성 추가 
- presigned URL 생성 후 6시간 동안 유효하도록 설정

khoon9 committed:
chore: yaml 파일 수정

- aws s3 통신에 필요한 프로퍼티 설정 완료
- 개발 환경에 맞춰 h2 메모리만 사용하도록 설정

khoon9 committed:
chore: 의존성 라이브러리 추가

- aws s3 통신을 위한 의존성 라이브러리 추가

khoon9 committed:
chore: aws s3 통신 설정

- 자격 증명 공급 설정

khoon9 committed:
feat: entity 추가 및 수정

- 프로필에 포스트 형태로 함께 소개될 Post entity 추가
- Post와의 다대일 관계 Profile entity에 반영
- Profile update을 위한 set 메소드 일부 추가

khoon9 committed:
feat: aws s3 통신 기반 이미지 서비스 구현

- 이미지 파일 upload, delete 구현
- presigned URL 기반 클라이언트 이미지 통신 구조 구현

khoon9 committed:
feat: 포스트 서비스 구현

- create request dto에 기반하여 포스트 생성 구현
- 포스트 조회시, 포스트에 속한 이미지의 presigned URL을 생성하여  image key 대신 반환 수행하도록 구현
- post 삭제 요청 시, 조회된 post entity 객체와 저장된 s3 이미지 제거 수행
- profile과의 다대일 관계를 고려하여, 여러 포스트 entity 객체를 리스트로 받아 presigned URL 발급 후 PostResponseDto 리스트에 담아 반환하는 메소드 구현

khoon9 committed:
feat: 포스트 API 추가

- 포스트 생성, 조회, 제거 API 추가

khoon9 committed:
chore: 서비스 및 예외 메시지 상수 추가

- 서비스 고유 설정(s3 bucket 내의 폴더 이름, presigned URL 유효 시간) 상수 추가
- 예외 메시지 상수 추가

khoon9 committed:
feat: 프로필 API 추가

- 프로필 수정 API 추가
- 프로필 수정 서비스 코드 구현

f-lab-green commented on src/main/java/flab/transtalk/config/AwsConfig.java:
```diff
@@ -0,0 +1,33 @@
+    public S3Presigner s3Presigner(){
+        return S3Presigner.builder()
+                .region(Region.of(region))
+                .credentialsProvider(DefaultCredentialsProvider.create())
```
DefaultCredentialsProvider.create()를 매번 생성하고 있네요. 여기서 큰 문제는 없겠지만 습관상 공통으로 사용할 것은 한 번만 생성하게 하는 습관이 좋아요.

khoon9 replied:
넵 ! 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/domain/Post.java:
```diff
@@ -0,0 +1,41 @@
+    private Long id;
+
+    @Column
+    @Builder.Default
```
👍

f-lab-green commented on src/main/java/flab/transtalk/user/domain/Post.java:
```diff
@@ -0,0 +1,41 @@
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class Post {
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE)
```
시퀀스를 사용한 이유가 있으신가요?

khoon9 replied:
시퀀스를 사용할 경우 JPA가 엔티티를 저장하기 전에 미리 시퀀스로 ID를 받아와서 IDENTITY 보다 비교적 더욱 향상된 퍼포먼스를 보여줄 수 있다고 알고 있습니다. 이에 따라, 배포 환경에서는 시퀀스를 지원하는 PostgreSQL을 염두에 두고 있습니다 !

f-lab-green commented on src/main/java/flab/transtalk/user/service/post/ImageService.java:
```diff
@@ -0,0 +1,60 @@
+    // image key 생성 및 이미지 업로드
+    public String uploadImageFile(MultipartFile file) throws IOException {
+        // image key 구성: post image folder 이름 + UUID + LocalDate + 확장자
+        String extension = ".jpg";
```
jpg로 고정되고 있는데 컨텐츠 타입에 따라 결정하는건 어떨까요.

khoon9 replied:
클라이언트에서 확장자를 확정짓는 작업을 먼저 수행한 후 서버로 전송하는 구조를 고려중에 있었습니다! 추후 관련해서 변경사항이 발생할 경우 컨텐츠 타입에 따라 결정할지 여부 또한 함께 고려하도록하겠습니다.

f-lab-green replied:
클라이언트에서 확장자를 결정하면 위변조가 가능해지지 않을까요?

khoon9 replied:
아하..! 말씀 주신 내용 반영하여, 업로드 이미지에 대해 MIME 타입 기반 확장자 판별 과정을 수행한 후 jpg 또는 png에 한하여 컨텐츠 타입에 따라 결정되도록 구현하였습니다 !!

f-lab-green commented on src/main/java/flab/transtalk/user/service/post/ImageService.java:
```diff
@@ -0,0 +1,60 @@
+
+    // s3 단일 image 삭제
+    public void deleteImageFile(String imageKey) {
+        s3Client.deleteObject(b -> b.bucket(bucket).key(imageKey));
```
삭제 대상이 없는 경우에 대한 예외처리도 있으면 좋을 것 같아요.

khoon9 replied:
넵. 기존 NotFoundException가 id 뿐만아니라 입력받은 내용을 포괄하여 예외를 처리하도록 구성한 후, 이미지 삭제 대상이 없는 경우에 대한 예외 처리를 추가하여 반영하였습니다 !

f-lab-green commented on src/main/java/flab/transtalk/user/service/post/ImageService.java:
```diff
@@ -0,0 +1,60 @@
+        s3Client.deleteObject(b -> b.bucket(bucket).key(imageKey));
+    }
+
+    public String generatePresignedUrl(String imageKey, Duration ttl){
```
지금은 괜찮지만 무조건 생성될 거라 유효성 검사가 나중에 추가되면 좋을 것 같네요.

khoon9 replied:
넵. 대상 이미지가 존재하지 않으면 앞서서 예외가 발생하도록하여 반영하였습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/dto/req/PostCreateRequestDto.java:
```diff
@@ -0,0 +1,13 @@
+
+@Value
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
```
여기서 이 친구는 필수가 아니여도 될 것 같아요. 물론 상관 없습니다.

khoon9 replied:
create 요청 dto에 대해선 생성자에 대한 접근 제한자를 비교적 엄격하지 않게 관리해도 괜찮다는 말씀이실까요?

f-lab-green replied:
아마 제 기억으로 필드가 3개뿐이라 굳이 필요 없다는 코멘트를 남겼던 것 같네요.

khoon9 replied:
넵 ! 이 부분 반영했습니다

f-lab-green commented on src/main/java/flab/transtalk/user/dto/res/PostResponseDto.java:
```diff
@@ -0,0 +1,15 @@
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class PostResponseDto {
+    private Long Id;
```
컨벤션 오류에요. Id -> id

khoon9 replied:
넵 수정했습니다. 감사합니다 !

f-lab-green commented on src/main/java/flab/transtalk/user/service/post/PostService.java:
```diff
@@ -0,0 +1,91 @@
+        post.setProfile(profile);
+
+        Post saved = postRepository.save(post);
+        String presignedUrl = imageService.generatePresignedUrl(generatedImageKey, Duration.ofHours(ServiceConfigConstants.PRESIGNED_URL_DURATION_HOUL));
```
HOUL -> HOUR ?

khoon9 replied:
헉 이 부분 오타 전부 수정하겠습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/service/post/PostService.java:
```diff
@@ -0,0 +1,91 @@
+
+    public List<PostResponseDto> getPosts(List<Post> posts) {
+        if (posts == null){
+            return null;
```
emptyList는 어떨까요?

khoon9 replied:
넵 ! 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/config/ServiceConfigConstants.java:
```diff
@@ -0,0 +1,6 @@
+
+public class ServiceConfigConstants {
+    public static final String S3_POST_IMAGE_FOLDER_NAME = "posts/";
+    public static final long PRESIGNED_URL_DURATION_HOUL = 6;
```
ㅋㅋㅋ 이거 커밋은 나중에 하셨나보군요. 인과관계상 앞선 커밋의 앞에 와야 할 것 같아요.

khoon9 replied:
넵 ㅎㅎ.. 제가 잠시 순서를 햇갈렸나봐요 ㅋㅋㅋㅋㅋ 앞으로는 인과관계에 더욱 유의하겠습니다 !

f-lab-green commented on src/main/java/flab/transtalk/user/controller/PostController.java:
```diff
@@ -0,0 +1,47 @@
+    public ResponseEntity<PostResponseDto> createPost(
+                @RequestParam("briefContext") String briefContext,
+                @RequestParam("imageFile") MultipartFile imageFile,
+                @RequestParam("profileId") Long profileId) throws Exception {
```
예외가 너무 포괄적이에요. 핸들러가 있다면 상관은 없겠지만요.

khoon9 replied:
넵 ! IOException에 대한 예외처리를 서비스 레이어에서 수행하도록 수정하여 반영하였습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/controller/ProfileController.java:
```diff
@@ -0,0 +1,28 @@
+
+    private final ProfileService profileService;
+
+    @PutMapping
```
PUT /api/profile/{profileId}
이 구조면 좀 더 명확할 것 같아요.

f-lab-green commented on src/main/java/flab/transtalk/user/dto/req/ProfileUpdateRequestDto.java:
```diff
@@ -0,0 +1,25 @@
+    String selfIntroduction;
+    LanguageSelection language;
+
+    public Profile toEntity(){
```
1. 필요 없다면 만들지 않는 것이 좋아요 -> YAGNI
2. 이건 업데이트이네 Profile이 새로 생성되고 있어서 사용할 때 주의를 해야할 것 같아요.

f-lab-green commented on src/main/java/flab/transtalk/user/service/profile/ProfileService.java:
```diff
@@ -0,0 +1,51 @@
+            profile.setLanguage(dto.getLanguage());
+        }
+
+        Profile updated = profileRepository.save(profile);
```
이 메서드에 Transactional을 추가하고 save 명시는 제거하는게 나을 것 같아요.

khoon9 replied:
넵 반영했습니다.

f-lab-green commented on src/main/java/flab/transtalk/user/service/profile/ProfileService.java:
```diff
@@ -0,0 +1,51 @@
+        return ProfileResponseDto.from(updated);
+    }
+
+    @Transactional
```
읽기 전용이면 readOnly = true를 붙이면 좋을 것 같네요.

khoon9 replied:
넵 반영했습니다 !

khoon9 committed:
fix: 자격 인증 공급 객체 생성 코드 수정

- DefaultCredentialsProvider.create() 한 번만 수행되도록 수정

khoon9 committed:
chore: 예외 처리 수정

- 기존 사용자 정의 NotFoundException 클래스를 수정함
- input만 예외에 포함한 경우와 input과 output 둘다 예외에 포함한 경우를 나누어 동작하도록 구성함

khoon9 committed:
chore: 이미지 삭제 예외 처리 추가

- S3 버킷 내의 이미지 삭제를 시도할 때, 삭제하고자 하는 이미지가 존재하지 않으면 예외가 발생하도록 구성함
- Presigned URL 발급을 시도할 때, 대상 이미지가 존재하는지 먼저 확인하여 존재하지 않으면 예외가 발생하도록 구현함

khoon9 committed:
chore: 오타 수정

- HOUL -> HOUR
- Id -> id

khoon9 committed:
chore: list 반환 형식 수정

 - return null  -> return 빈 리스트로 변경

khoon9 committed:
chore: 예외 범위 명시

- Exception을 IOException으로 명시하여 처리

khoon9 committed:
chore: 프로필 수정 요청 url 변경

- profileId를 PathVariable로 받도록 수정

khoon9 committed:
chore: 프로필 수정 및 조회 성능 개선

- 불필요한 save 호출 제거로 프로필 수정 성능 개선
- 조회 전용 트랜잭션 설정으로 프로필 조회 성능 개선

khoon9 committed:
feat: 업로드 이미지 MIME 타입 기반 확장자 판별 추가

- 업로드된 이미지의 실제 콘텐츠 타입을 판별하고, 허락된 확장자에 한하여 업로드되도록 수정
- Apache Tika을 사용하여 MIME 타입 기반 확장자 결정 수행

khoon9 committed:
chore: 어노테이션 수정

- @AllArgsConstructor(access = AccessLevel.PRIVATE) 코드 제거

khoon9 committed:
fix: 스크립트 환경변수 설정 추가

khoon9 committed:
fix: gradle wrapper 버전 수정

- Gradle 8.12.1 -> 8.14.1

khoon9 committed:
fix: 스크립트 java 명시 버전 및 명령어 수정

khoon9 committed:
chore: 불필요한 코드 정리

- unused import 제거
- empty statement 제거

khoon9 committed:
chore: 서비스 설정 클래스 수정

- 상수만 가지는 유틸리티 클래스이므로 인스터스화 안 되도록 설정

khoon9 committed:
chore: 예외 클래스 수정

- 필드를 final로 선언하여 불변성을 보장하도록 수정
